src/plugins/matrixops/concatenate.cpp

```diff
@@ -0,0 +1,835 @@
+#endif
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const concatenate::match_data = {
+        hpx::util::make_tuple("concatenate",
+            std::vector<std::string>{"concatenate(_1)", "concatenate(_1, _2)"},
+            &create_concatenate, &create_primitive<concatenate>, R"(
+            ar, axis
+            Args:
+
+                ar (array) : sequence of array_like
+                axis (optional, int) : the axis along which the arrays will be joined,
+                    default is 0, if axis is None arrays are flattened
+                    before use
+
+            Returns:
+
+            The joined sequence of arrays along an existing axis."
+
+            )")};
+
+    ///////////////////////////////////////////////////////////////////////////
+    concatenate::concatenate(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    std::size_t concatenate::get_vec_size(
+        primitive_arguments_type const& args) const
+    {
+        std::size_t vec_size = 0;
+
+        for (std::size_t i = 0; i != args.size(); ++i)
+        {
+            std::size_t num_dims =
+                extract_numeric_value_dimension(args[i], name_, codename_);
+
+            if (num_dims != 1)
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "phylanx::execution_tree::primitives::concatenate::get_"
+                    "tensor_size",
+                    generate_error_message("the concatenate primitive requires "
+                                           "for all input arrays to "
+                                           "have the same dimension"));
+
+            if (num_dims == 0)
+            {
+                ++vec_size;
+            }
+            else
+            {
+                vec_size += extract_numeric_value_dimensions(
+                    args[i], name_, codename_)[0];
+            }
+        }
+
+        return vec_size;
+    }
+
+    template <typename T>
+    primitive_argument_type concatenate::concatenate_flatten1d(
+        primitive_arguments_type&& args) const
+    {
+        blaze::DynamicVector<T> result(get_vec_size(args));
+
+        auto iter = result.begin();
+        for (auto&& arg : args)
+        {
+            auto&& val = extract_node_data<T>(std::move(arg));
+            std::size_t num_d = val.num_dimensions();
+            if (num_d == 0)
+            {
+                *iter++ = val.scalar();
+            }
+            else
+            {
+                std::copy(val.vector().begin(), val.vector().end(), iter);
+                iter += val.size();
```

Comment:
`std::copy` returns an iterator that points to the first element after the one assigned last (see https://en.cppreference.com/w/cpp/algorithm/copy). So the above could be simplified as:
```
iter = std::copy(val.vector().begin(), val.vector().end(), iter);
```

Reply:
Also, you might want to use `std::move` instead of `std::copy` here (https://en.cppreference.com/w/cpp/algorithm/move).